File: Week2/Week2_ProblemB.cpp
#include <iostream>
using namespace std;

float d, p, u, v;

void buildResult(int caseNo) {
    
    float max = d / p;				//doesn't fit
    float min = (d - (v - u)) / p;	//fit
    float mid = (max + min) / 2;
    
    float start = 0;
    int factor = 0;
    
    while (max - min > 0.0001f) {
        
        factor = (int) (v / mid);
        start += (v - factor * mid);
        start += (mid * (p + (factor - (int) (u / mid) - 1)));
        
        if (start > d)
            max = mid;
        else
            min = mid;
        
        mid = (max + min) / 2;
        start = 0;
    }
    
    cout << "Case #" << caseNo << ": " << mid << endl;
}


int main(int argc, const char * argv[]) {
    
    std::ios::sync_with_stdio(false);
    
    int n;
    
    cin >> n;
    
    for (int i = 1; i <= n; i++) {
        cin >> d >> p >> u >> v;
        p--;
        buildResult(i);
    }
    
    return 0;
}

File: Week2/Week2_ProblemB.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Week2_ProblemB {
	static double d, p, u, v;
	
	public static void main(String[] args) throws NumberFormatException, IOException {
		
		InputStreamReader r = new InputStreamReader ( System . in );
		BufferedReader in = new BufferedReader ( r );
		int testCasesCount = Integer.parseInt(in.readLine());
		
		for (int i = 1; i <= testCasesCount; i++) {
			String line = in.readLine();
			String[] values = line.split(" ");
			d = Integer.parseInt(values[0]);
			p = Integer.parseInt(values[1]) - 1;
			u = Integer.parseInt(values[2]);
			v = Integer.parseInt(values[3]);
					
			buildResult(i);
		}
		
	}
	
	public static void buildResult(int caseNo) {
		
		double max = d / p;				//doesn't fit
		double min = (d - (v - u)) / p;	//fit
		double mid = (max + min) / 2;
		
		double start = 0;
		int factor;
		
		while (max - min > 0.001f) {
			
			factor = (int) (v / mid);
			start += (v - factor * mid);
			start += (mid * (p + (factor - (int) (u / mid) - 1)));
			
			if (start > d)
				max = mid;
			else
				min = mid;
			
			mid = (max + min) / 2;
			start = 0;
		}
		
		System.out.println("Case #" + caseNo + ": " + mid);
	}
}
